File: lintcode/129-Rehashing/solution.cpp
/**
 * Definition of ListNode
 * class ListNode {
 * public:
 *     int val;
 *     ListNode *next;
 *     ListNode(int val) {
 *         this->val = val;
 *         this->next = NULL;
 *     }
 * }
 */
class Solution {
public:
    /**
     * @param hashTable: A list of The first node of linked list
     * @return: A list of The first node of linked list which have twice size
     */    
    vector<ListNode*> rehashing(vector<ListNode*> hashTable) {
        // double table size
        int old_size = hashTable.size();
        for (int i = 0; i < old_size; i++) {
            hashTable.push_back(nullptr);
        }
        
        // rehashing
        int new_size = old_size * 2;
        for (int i = 0; i < old_size; i++) {
            if (hashTable[i] == nullptr) {
                continue;
            }
            
            // the first node might be changed
            ListNode dummy(0);
            dummy.next = hashTable[i];
            ListNode *prev = &dummy;
            while (prev->next != nullptr) {
                int new_index = hashKey(prev->next->val, new_size);
                if (new_index != i) {
                    // move current node to new one
                    if (hashTable[new_index] == nullptr) {
                        hashTable[new_index] = prev->next;
                        prev->next = prev->next->next;
                        hashTable[new_index]->next = nullptr;
                    } else {
                        ListNode *tail = hashTable[new_index];
                        while (tail->next != nullptr) {
                            tail = tail->next;
                        }
                        tail->next = prev->next;
                        prev->next = prev->next->next;
                        tail->next->next = nullptr;
                    }
                } else {
                    prev = prev->next;
                }
            }
            hashTable[i] = dummy.next;
        }
        
        return hashTable;
    }
private:
    int hashKey(int key, int table_size) {
        return (key % table_size + table_size) % table_size;
    }
};

File: lintcode/129-Rehashing/solution.java
/**
 * Definition for ListNode
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    /**
     * @param hashTable: A list of The first node of linked list
     * @return: A list of The first node of linked list which have twice size
     */    
    public ListNode[] rehashing(ListNode[] hashTable) {
        if (hashTable == null || hashTable.length == 0) {
            return hashTable;
        }
        
        ListNode[] newHashTable = new ListNode[hashTable.length * 2];

        int len = hashTable.length;
        for (int i = 0; i < len; i++) {
            if (hashTable[i] == null) {
                continue;
            }
            // move current list's nodes to new position
            ListNode dummy = new ListNode(0);
            dummy.next = hashTable[i];
            ListNode prev = dummy;
            ListNode curr = dummy.next;
            while (curr != null) {
                int newKey = hashKey(curr.val, len * 2);
                if (newKey != i) {
                    prev.next = curr.next;
                    if (newHashTable[newKey] == null) {
                        newHashTable[newKey] = curr;
                        curr.next = null;
                    } else {
                        ListNode tail = newHashTable[newKey];
                        while (tail.next != null) {
                            tail = tail.next;
                        }
                        tail.next = curr;
                        curr.next = null;
                    }
                    curr = prev.next;
                } else {
                    prev = prev.next;
                    curr = curr.next;
                }
            }
            newHashTable[i] = dummy.next;
        }
        
        return newHashTable;
    }
    
    private int hashKey(int key, int tableSize) {
        return (key % tableSize + tableSize) % tableSize;
    }
}
